brave/brave-ios
# Fix #1119: Adds Bookmarklet support with validation.

- Adds Bookmarklet support with validation.
- Fixes https://github.com/brave/brave-ios/issues/1119
- Allow JS to run in the URL bar (can be disabled and Bookmarklets will still work - or make it into a preference).
- Security Review: https://github.com/brave/internal/issues/637

When adding a bookmarklet, if you enter into the URL field:

`"javascript:"` and some JS code right after it, it will save as a `Bookmarklet`.
The next time you visit any page, you can execute the bookmarklet and it will modify the web-view/page/content accordingly.

IE (Bookmark):

title: `"view-source"`
url: 
```javascript
javascript:(function(){var a=window.open('about:blank').document;a.write('<!DOCTYPE html><html><head><title>Source of '+location.href+'</title><meta name="viewport" content="width=device-width" /></head><body></body></html>');a.close();var b=a.body.appendChild(a.createElement('pre'));b.style.overflow='auto';b.style.whiteSpace='pre-wrap';b.appendChild(a.createTextNode(document.documentElement.innerHTML))})();
```

Save it as a bookmark. Go to lets say `youtube` for example and tap on bookmarks -> view-source.. It will now execute the above code and you'll be able to view the source-code of youtube for example..

Another one is that you can run JS in the URL bar by typing in the URL bar like:
```javascript
javascript:location.href="https://stackoverflow.com";
```

Note: Brave does not support `view-source` URI natively.. but it does through bookmarklet support if the PR goes through :)

Note: Not sure if it should be integrated like this or in a separate menu or something. No idea, but it's here.

## Submitter Checklist:



## Test Plan:



### Screenshots:




## Reviewer Checklist:



**TEST PLAN:**
**Test Plan:**

1. Add a new bookmark.
2. Change the url to `javascript:function(){}`
3. It should fail to save.

-

1. Add a new bookmark.
2. Change the url to `javascript:`
3. Should fail to save.

-

1. Add a new bookmark.
2. Change the url to `https://brave.com`
3. It MUST save successfully!
4. Open the bookmark and it should go to the correct url.
5. Any current bookmarks MUST work as well.

-

1. Add a new bookmark.
2. Change the url to `javascript:(function(){ alert("hello"); })()`
3. Save it successfully.
4. Go to any website and open your bookmark.
5. Should show a popup saying `“Hello”`.

ONLY if all the above steps for EVERY case works, this ticket can be marked resolved. If one has failed, we **MUST** file a new bug or message me, as bookmarks are a sensitive thing for users and we don't want to lose any functionality for them.
